src/plugins/discover/public/kibana_services.ts

```diff
@@ -60,10 +60,23 @@ export const [getDocViewsRegistry, setDocViewsRegistry] = createGetterSetter<Doc
   'DocViewsRegistry'
 );
 /**
- * Makes sure discover and context are using one instance of history
+ * Makes sure discover and context are using one instance of history.
  */
 export const getHistory = _.once(() => createHashHistory());
 
+/**
+ * Discover currently uses two `history` instances: one from Kibana Platform and
+ * another from `history` package. Below function is used every time Discover
+ * app is loaded to synchronize both instances.
+ *
+ * This helper is temporary until https://github.com/elastic/kibana/issues/65161 is resolved.
+ */
+export const syncHistoryLocations = () => {
+  const h = getHistory();
+  Object.assign(h.location, createHashHistory().location);
+  return h;
+};
+
```

Comment:
👍 for making this work

Reply:
Kudos to @Dosant for coming up with this fix. But Discover still needs to transition to a single `history` instance, for example this fix synchronizes both instances when the app mounts, not sure what happens afterwards, if state is synced to bot instances or they both can diverge again.